sklearn/mixture/bayesianmixture.py

```diff
@@ -0,0 +1,477 @@
+            "spherical": self._estimate_lambda_spherical
+        }
+        return estimate_lambda_functions[self.covariance_type](X, nk, xk, Sk)
+
+    def _e_step(self, X):
+        log_prob_comp, log_prob, resp = \
+            self._estimate_log_probabilities_responsibilities(X)
+        lower_bound = self._lower_bound(log_prob_comp, resp)
+        return lower_bound, resp
+
+    def _m_step(self, X, nk, xk, Sk):
+        self.weight_alpha_ = self._estimate_weights(X, nk, xk, Sk)
+        self.mu_beta_, self.mu_m_ = self._estimate_means(X, nk, xk, Sk)
+        self.lambda_nu_, self.lambda_inv_W_ = self._estimate_lambda(
+            X, nk, xk, Sk)
+
+    # e step
+    def _estimate_weighted_log_probabilities(self, X):
+        # save ln_weights for computing lower bound
+        log_weights = self._estimate_log_weights()
+        _estimate_log_X = {
+            "full": self._estimate_log_X_full,
+            "tied": self._estimate_log_X_tied,
+            "diag": self._estimate_log_X_diag,
+            "spherical": self._estimate_log_X_spherical
+        }
+        # save ln_lambda, inv_W_chol for computing lower bound
+        log_X = _estimate_log_X[self.precision_type](X)
+        return log_weights + log_X
+
+    def _estimate_log_weights(self):
+        # first item in Equation 3.8
+        log_weights = (digamma(self.weight_alpha_) -
+                digamma(np.sum(self.weight_alpha_)))
+        self.log_weights = log_weights
+        return log_weights
+
+    def _estimate_log_X_full(self, X):
+        # second item in Equation 3.8
+        n_samples = X.shape[0]
+        n_features = self.n_features
+        ln_W_digamma = np.arange(1, self.n_features + 1)
+        log_X = np.empty((n_samples, self.n_components))
+        log_lambda = np.empty((self.n_components, ))
+        inv_W_chol = np.empty(self.lambda_inv_W_.shape)
+        for k, (beta, m, nu, inv_W) in enumerate(
+                zip(self.mu_beta_, self.mu_m_,
+                    self.lambda_nu_, self.lambda_inv_W_)):
+            try:
+                inv_W_chol[k] = linalg.cholesky(inv_W, lower=True)
+            except linalg.LinAlgError:
+                raise ValueError("'lambda_inv_W_' must be symmetric, "
+                                 "positive-definite")
+            ln_inv_W_det = 2. * np.sum(np.log(np.diagonal(inv_W_chol[k])))
+            log_lambda[k] = (np.sum(digamma(.5 * (nu + 1 - ln_W_digamma))) +
+                             n_features * np.log(2) - ln_inv_W_det)
+
+            W_sol = linalg.solve_triangular(inv_W_chol[k], (X - m).T,
+                                            lower=True).T
+            mahala_dist = np.sum(np.square(W_sol), axis=1)
+            log_X[:, k] = - .5 * (- log_lambda[k] +
+                                  n_features / beta + nu * mahala_dist)
+        log_X -= .5 * (n_features * np.log(2 * np.pi))
+        self.inv_W_chol = inv_W_chol
+        self.log_lambda = log_lambda
```

Comment:
Hi @ogrisel @amueller @vene @lesteve I didn't save resp and other data which could scale along with n_samples, but I did save other temporary data as instance attributes to compute lower bound when doing e_step. I don't know this is good or bad style.  Any comment?